File: Leetcode/java/medium/1302_Deepest Leaves Sum.java
/*
 * @lc app=leetcode id=1302 lang=java
 *
 * [1302] Deepest Leaves Sum
 */

// @lcpr-template-start
import java.util.*;

class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
// @lcpr-template-end
// @lc code=start

class Solution1 {

    private int maxDepth = 0;
    private int maxDepthSum = 0;

    public int deepestLeavesSum(TreeNode root) {
        dfs(root, 0);
        return maxDepthSum;
    }

    private void dfs(TreeNode node, int depth) {
        if (depth > maxDepth) {
            maxDepth = depth;
            maxDepthSum = node.val;
        } else if (depth == maxDepth) {
            maxDepthSum += node.val;
        }
        if (node.left != null) {
            dfs(node.left, depth + 1);
        }
        if (node.right != null) {
            dfs(node.right, depth + 1);
        }
    }
}

class Solution2 {

    public int deepestLeavesSum(TreeNode root) {
        int maxDepthSum = 0;
        Queue<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        while (!q.isEmpty()) {
            maxDepthSum = 0;
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                TreeNode node = q.poll();
                maxDepthSum += node.val;
                if (node.left != null) {
                    q.add(node.left);
                }
                if (node.right != null) {
                    q.add(node.right);
                }
            }
        }
        return maxDepthSum;
    }
}

class Solution extends Solution1 {}
// class Solution extends Solution2 {}
// @lc code=end

File: Leetcode/python3/medium/1302_Deepest Leaves Sum.py
#
# @lc app=leetcode id=1302 lang=python3
#
# [1302] Deepest Leaves Sum
#


# @lcpr-template-start
from preImport import *
# @lcpr-template-end
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution1:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        max_depth = max_depth_sum = 0

        def dfs(node: TreeNode, depth: int) -> None:
            nonlocal max_depth, max_depth_sum
            if depth > max_depth:
                max_depth = depth
                max_depth_sum = node.val
            elif depth == max_depth:
                max_depth_sum += node.val
            if node.left is not None:
                dfs(node.left, depth + 1)
            if node.right is not None:
                dfs(node.right, depth + 1)

        dfs(root, 0)

        return max_depth_sum

class Solution2:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        q = deque([root])
        while q:
            max_depth_sum = 0
            for _ in range(len(q)):
                node = q.popleft()
                max_depth_sum += node.val
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
        return max_depth_sum


Solution = Solution1
# Solution = Solution2
# @lc code=end

